src/ffi/shared_library.rs

```diff
@@ -0,0 +1,251 @@
+/// [`run_query_iter`] that has not yet been freed by
+/// [`query_state_free`], or the [`Machine`] will enter an undefined state
+/// with unpredictable behavior.
+///
+/// # Arguments
+///
+/// * `machine` - A mutable reference to the [`Machine`] to load the module into.
+/// * `module_name` - A pointer to a null-terminated UTF-8 string representing
+/// Scryer Prolog module name.
+/// * `input` - A pointer to a null-terminated UTF-8 string representing Scryer Prolog code.
+///
+#[export_name = "scryer_consult_module_string"]
+pub unsafe extern "C" fn consult_module_string(
+    machine: &mut Machine,
+    module_name: *const c_char,
+    input: *const c_char,
+) {
+    let c_str: &CStr;
+    unsafe {
+        c_str = CStr::from_ptr(input);
+    }
+    let r_str = c_str.to_str().expect("Not a valid UTF-8 string");
+
+    let m_str: &CStr;
+    unsafe {
+        m_str = CStr::from_ptr(module_name);
+    }
+    let module_name = m_str.to_str().expect("Not a valid UTF-8 string");
+
+    machine.consult_module_string(&module_name, r_str.to_owned())
+}
+
+/// Greedily evaluate a prolog query, returning all results in a JSON-formatted string.
+/// Errors are printed to stderr and a null pointer is returned.
+///
+/// # Safety
+///
+/// * `machine` must be a valid [`Machine`] pointer created with [`machine_new`]
+/// that has not yet been freed.
+/// * There must be no existing [`QueryState`] for this [`Machine`] started by
+/// [`run_query_iter`] that has not yet been freed with
+/// [`query_state_free`] or the [`Machine`] state will enter an undefined
+/// configuration with unpredictable results.
+/// * it is the responsibility of the caller to deallocate the pointer returned by
+/// this function with [`scyer_free_c_string`] in order to avoid memory leaks.
+///
+///
+/// # Returns
+/// - Returns a pointer to a JSON formatted, null terminated UTF-8 string
+///
+/// # Response Format
+/// ```json
+/// // if result is a binding
+///
+/// // current limitation is that only concrete (equality) bindings are returned,
+/// // residual goals not yet supported.
+/// {
+///   "status": "ok",
+///   "result": [{ ... }],
+/// }
+///
+/// // if result is a boolean goal
+///
+/// {
+///   "status": "ok",
+///   "result": boolean
+/// }
+#[export_name = "scryer_run_query"]
+pub unsafe extern "C" fn run_query(machine: &mut Machine, input: *const c_char) -> *mut c_char {
+    let c_string;
+    let r_str;
+    unsafe {
+        c_string = CStr::from_ptr(input);
+    };
+    r_str = c_string.to_str().expect("Not a valid UTF-8 string");
+    let query_resolution = machine.run_query(r_str.to_owned());
+
+    let output_string: String = match query_resolution {
+        Ok(query_resolution_line) => {
+            let value: Value = serde_json::from_str(&format!("{}", query_resolution_line)).unwrap();
+            json!( {
+                "status": "ok",
+                "result": value
+            })
+            .to_string()
```

Comment:
With this being rebased onto #2493 I would expect that this can be simplified to

```suggestion
            json!({
                "status": "ok",
                "result": query_resolution_line
            })
            .to_string()
```